File: app/frontend/types/dispute.ts
import { User } from './auth';
import { SelectOption } from './profile';

export interface Dispute {
  id?: number;
  job_id: number;
  reporter_id: number;
  reported_user_id: number;
  reason: string;
  description: string;
  status?: 'pending' | 'resolved' | 'escalated';
  notes?: string; // Added notes property
  created_at?: string;
  updated_at?: string;
  job?: { id: number; title: string }; // Added job property
  reporter?: User;
  reported_user?: User;
  errors?: Record<string, string[]>;
}

export interface DisputeFormProps {
  job: { id: number; title: string };
  dispute: Dispute;
  errors: Record<string, string[]>;
  reportedUserOptions: SelectOption[];
}

export interface AdminDisputeIndexProps {
  disputes: Dispute[];
}

export interface AdminDisputeShowProps {
  dispute: Dispute;
}

File: app/models/dispute.rb
# frozen_string_literal: true

class Dispute < ApplicationRecord
  belongs_to :job
  belongs_to :reporter, class_name: "User", foreign_key: "reporter_id"
  belongs_to :reported_user, class_name: "User", foreign_key: "reported_user_id"

  enum :status, {
    pending: 0,
    resolved: 1,
    escalated: 2
  }, default: :pending, validate: true

  validates :reason, presence: true, length: {maximum: 255}
  validates :description, presence: true, length: {maximum: 1000}
  validates :status, presence: true
  validates :job_id, presence: true
  validates :reporter_id, presence: true
  validates :reported_user_id, presence: true
  validate :cannot_report_self

  private

  def cannot_report_self
    if reporter_id == reported_user_id
      errors.add(:reported_user_id, "cannot be the same as the reporter")
    end
  end
end
